ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ContractsReader.scala

```diff
@@ -6,47 +6,147 @@ package com.daml.platform.store.appendonlydao.events
+          .as(
+            (contractId("contract_id") ~ flatEventWitnessesColumn("flat_event_witnesses")).map {
+              case cId ~ stakeholders => KeyAssigned(cId, stakeholders)
+            }.singleOpt
+          )
+          .getOrElse(KeyUnassigned)
+      },
+    )
 
-  private val contractWithoutValueRowParser: RowParser[String] =
-    str("template_id")
+  override def lookupContractState(contractId: ContractId, before: Long)(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[ContractState]] =
+    Timed.future(
+      metrics.daml.index.db.lookupContractState,
+      dispatcher
+        .executeSql(metrics.daml.index.db.lookupContractStateDbMetrics) { implicit connection =>
+          SQL"""
+           SELECT
+             template_id,
+             flat_event_witnesses,
+             create_argument,
+             create_argument_compression,
+             event_kind,
+             ledger_effective_time
+           FROM participant_events
+           WHERE
+             contract_id = $contractId
+             AND event_sequential_id <= $before
+             AND (event_kind = #$EventKindCreated OR event_kind = #$EventKindArchived)
+           ORDER BY event_sequential_id DESC
+           LIMIT 1;
+           """
+            .as(fullDetailsContractRowParser.singleOpt)
+        }
+        .map(_.map {
+          case (
+                templateId,
+                stakeholders,
+                createArgument,
+                createArgumentCompression,
+                EventKindCreated,
+                maybeCreateLedgerEffectiveTime,
+              ) =>
+            val contract = toContract(
+              contractId = contractId,
+              templateId =
+                assertPresent(templateId)("template_id must be present for a create event"),
+              createArgument =
+                assertPresent(createArgument)("create_argument must be present for a create event"),
+              createArgumentCompression =
+                Compression.Algorithm.assertLookup(createArgumentCompression),
+              decompressionTimer =
+                metrics.daml.index.db.lookupContractStateDbMetrics.compressionTimer,
+              deserializationTimer =
+                metrics.daml.index.db.lookupContractStateDbMetrics.translationTimer,
+            )
+            ActiveContract(
+              contract,
+              stakeholders,
+              assertPresent(maybeCreateLedgerEffectiveTime)(
+                "ledger_effective_time must be present for a create event"
+              ),
+            )
+          case (_, stakeholders, _, _, EventKindArchived, _) => ArchivedContract(stakeholders)
+          case (_, _, _, _, kind, _) => throw ContractsReaderError(s"Unexpected event kind $kind")
+        }),
+    )
 
   /** Lookup of a contract in the case the contract value is not already known */
-  private def lookupActiveContractAndLoadArgument(
+  override def lookupActiveContractAndLoadArgument(
```

Comment:
set of metrics used by this method is different between `dao` and `appendonly`

Reply:
Updated. After our offline discussion, for compatibility reasons, I've reused the lookupActiveContract and lookupContractKey metrics for the new methods as well. To be revisited though